src/v/datalake/translation_task.h

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+#pragma once
+
+#include "base/outcome.h"
+#include "datalake/cloud_data_io.h"
+#include "datalake/coordinator/translated_offset_range.h"
+#include "datalake/data_writer_interface.h"
+#include "datalake/fwd.h"
+#include "model/record_batch_reader.h"
+#include "utils/lazy_abort_source.h"
+#include "utils/retry_chain_node.h"
+
+namespace datalake {
+/**
+ * An abstraction representing a task of consuming data, translating them to
+ * parquet and uploading to the object store
+ */
+class translation_task {
+public:
+    explicit translation_task(cloud_data_io& uploader);
+    enum class errc {
+        file_io_error,
+        cloud_io_error,
+    };
+    /**
+     * Executes the translation and uploads files to the object store. The tasks
+     * accepts an abort source indicating when the upload retires should be
+     * stopped and a root retry chain node.
+     */
+    ss::future<checked<coordinator::translated_offset_range, errc>> translate(
+      std::unique_ptr<data_writer_factory> writer_factory,
+      model::record_batch_reader reader,
```

Comment:
should this be kafka::read_committed_reader ?

Reply:
i made the interface generic so it is easier to test, the `kafka::read_committed_reader` is an implementation of `model::record_batch_reader`